Multi Monitor Support

setzer22 opened the pull request:
As we discussed in issue #46, these are the changes that improve multi-monitor support. I've been testing this for a couple of days and it seems to behave now in the same way as i3bar.

As you can see, the new approach of "re-query everything on every workspace event" made some of the old events not make sense anymore. I did my best to try to adapt the old event codes to this refactor:

- I've kept all events in `i3wm-delegate`, but now they just call `init_workspaces` (the old structure is freed, if existing) and then call the most appropiate callback (or when none made sense, I just used `on_workspace_created` and noted that in a comment).
- I've replaced all callbacks in `i3w-plugin` by `on_workspace_changed`, so now all workspace events from i3 fire this function.

I added a new config flag called "Auto detect outputs", should be self explanatory. It's disabled by default. Feel free to change that if you think that's better.

I also tried to make the overall coding style match as closely as possible the already existing one. Please let me know if you would want me to improve something.

denesb commented:
Thanks, I will try to review and merge this during the course of the week.

denesb commented on panel-plugin/i3w-multi-monitor-utils.c:
```diff
@@ -0,0 +1,140 @@
+ *
+ * Returns: True iff c indicated connected display
+ */
+int
```
use `gboolean`

denesb commented on panel-plugin/i3w-multi-monitor-utils.c:
```diff
@@ -0,0 +1,140 @@
+                int y = info->y;
+
+                char* name_dest = malloc(strlen(name)*sizeof(char));
+                strcpy(name_dest, name);
```
Why not write these directly into the output struct?

denesb commented on panel-plugin/i3w-multi-monitor-utils.c:
```diff
@@ -0,0 +1,140 @@
+                output->x = x;
+                output->y = y;
+
+                num_connected_outputs += 1;
```
++

denesb commented on panel-plugin/i3w-multi-monitor-utils.c:
```diff
@@ -0,0 +1,140 @@
+        outputs.num_outputs = num_connected_outputs;
+    }
+
+    outputs.num_outputs = num_connected_outputs;
```
Why is this done here again?

denesb commented on panel-plugin/i3w-multi-monitor-utils.c:
```diff
@@ -0,0 +1,140 @@
+void
+free_outputs(i3_workspaces_outputs_t outputs) {
+  free(outputs.outputs);
+}
```
There is no need for this indirection, you can call `free()` directly.

setzer22 replied:
You mean free(outputs.outputs) directly, right? I don't think free(outputs) would work, would it?

denesb replied:
I meant `free(outputs.outputs)` but on second thought this is better because if outputs changes only this function has to be updated (not all the callsites).

setzer22 replied:
That's right, I went back `free_outputs`.

denesb replied:
Sorry I promise I'll stop with the comments. :))

denesb commented on panel-plugin/i3w-plugin.c:
```diff
@@ -382,61 +398,75 @@ on_workspace_created(i3workspace *workspace, gpointer data)
     i3WorkspacesPlugin *i3_workspaces = (i3WorkspacesPlugin *) data;
+	if (!strncmp(mode, "default", 7))
+		gtk_label_set_text((GtkLabel *) i3_workspaces->mode_label, "");
+	else {
```
If you use brackets for the else, please use them for the if as well.

denesb commented on panel-plugin/i3w-plugin.c:
```diff
@@ -382,61 +398,75 @@ on_workspace_created(i3workspace *workspace, gpointer data)
+    int x, y;
+    GdkWindow* window = gtk_widget_get_window(i3_workspaces->ebox);
+    gdk_window_get_root_origin(window, &x, &y);
+    //printf("Widget window coordinates: x:%d, y:%d\n", x, y);
```
Debug stuff left in the code.

denesb commented on panel-plugin/i3wm-delegate.c:
```diff
@@ -767,6 +712,22 @@ on_mode_event(i3ipcConnection *conn, i3ipcGenericEvent *e, gpointer i3w) {
+/**
  * on_ipc_shutdown_proxy:
  * @connection: the ipc connection object
  * @i3wm - pointer to the window manager delegate struct
```
I like how much shorter this file is now.

setzer22 commented:
I have addressed your comments in the review. Please note my comment about `free` (did you mean to use `free(outputs.outputs)` directly?).

Thanks for the thorough review!

denesb commented on panel-plugin/i3w-multi-monitor-utils.c:
```diff
@@ -0,0 +1,119 @@
+            if (info) {
+                i3_workspaces_output_t* output = &outputs.outputs[num_connected_outputs];
+
+                output->name = malloc(strlen(output_info->name)*sizeof(char));
```
You need `strlen(output_info->name) + 1` as `strlen()` doesn't include the null terminator.

denesb commented:
Looks good, I just noticed one more thing. Yes I meant `free(outputs.outputs)`.

denesb commented on panel-plugin/i3w-multi-monitor-utils.h:
```diff
@@ -0,0 +1,37 @@
+
+i3_workspaces_outputs_t get_outputs();
+
+char* get_monitor_name_at(i3_workspaces_outputs_t outputs, int win_x, int win_y);
```
`const char*`
